File: Cpp/LRUCache.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <set>
#include <cassert>
using namespace std;

struct Node{
   Node* next;
   Node* prev;
   int value;
   int key;
   Node(Node* p, Node* n, int k, int val):prev(p),next(n),key(k),value(val){};
   Node(int k, int val):prev(NULL),next(NULL),key(k),value(val){};
};

class Cache{
   
   protected: 
   map<int,Node*> mp; //map the key to the node in the linked list
   int cp;  //capacity
   Node* tail; // double linked list tail pointer
   Node* head; // double linked list head pointer
   virtual void set(int, int) = 0; //set function
   virtual int get(int) = 0; //get function

};
class LRUCache : public Cache {
    public:
        LRUCache(int capacity) {
            cp = capacity;
            tail = nullptr;
            head = nullptr;
        }
        
        void moveToFront(Node* node) {
            if(node == head) {
                return;
            }
            
            if(node->prev) node->prev->next = node->next;
            if(node->next) node->next->prev = node->prev;
            
            if(node == tail) tail = node->prev;
            
            
            node->next = head;
            node->prev = nullptr;
            
            if(head) 
                head->prev = node;
            
            head = node;
            
            if(!tail) 
                tail = head;
        }
        
        int get(int key) {
            if(mp.find(key) == mp.end())
                return -1;
            
            Node* node = mp[key];
            moveToFront(node);
            return node->value;
        }
        
        void set(int key, int value) {
            if(mp.find(key) != mp.end()) {
                Node* node = mp[key];
                node->value = value;
                moveToFront(node);
            } else {
                if(mp.size() == cp) {
                    mp.erase(tail->key);
                    if(tail->prev) {
                        tail = tail->prev;
                        delete tail->next;
                        tail->next = nullptr;
                    } else {
                        delete tail;
                        tail = nullptr;
                        head = nullptr;
                    }
                }
                
                Node* newNode = new Node(key, value);
                newNode->next = head;
                if(head) head->prev = newNode;
                head = newNode;
                
                if(!tail) tail = head;
                
                mp[key] = newNode;
            }
        }
};
int main() {
   int n, capacity,i;
   cin >> n >> capacity;
   LRUCache l(capacity);
   for(i=0;i<n;i++) {
      string command;
      cin >> command;
      if(command == "get") {
         int key;
         cin >> key;
         cout << l.get(key) << endl;
      } 
      else if(command == "set") {
         int key, value;
         cin >> key >> value;
         l.set(key,value);
      }
   }
   return 0;
}

File: Languages/go/low-level-design-lld/problems/lru_cache.go
package main

import "fmt"

// LRU Cache with doubly-linked list + hash map.
// O(1) Get and Put operations.
// When capacity is exceeded, the least recently used item is evicted.

type node struct {
	key, val   int
	prev, next *node
}

type LRUCache struct {
	capacity int
	cache    map[int]*node
	head     *node // Most recently used (dummy)
	tail     *node // Least recently used (dummy)
}

func NewLRUCache(capacity int) *LRUCache {
	head := &node{}
	tail := &node{}
	head.next = tail
	tail.prev = head

	return &LRUCache{
		capacity: capacity,
		cache:    make(map[int]*node),
		head:     head,
		tail:     tail,
	}
}

// Get returns the value and marks it as recently used. Returns -1 if not found.
func (l *LRUCache) Get(key int) int {
	if n, ok := l.cache[key]; ok {
		l.remove(n)
		l.insertFront(n)
		return n.val
	}
	return -1
}

// Put inserts or updates a key-value pair. Evicts LRU item if at capacity.
func (l *LRUCache) Put(key, value int) {
	if n, ok := l.cache[key]; ok {
		n.val = value
		l.remove(n)
		l.insertFront(n)
		return
	}

	if len(l.cache) >= l.capacity {
		// Evict least recently used (node before tail)
		lru := l.tail.prev
		l.remove(lru)
		delete(l.cache, lru.key)
	}

	newNode := &node{key: key, val: value}
	l.cache[key] = newNode
	l.insertFront(newNode)
}

// remove detaches a node from the linked list.
func (l *LRUCache) remove(n *node) {
	n.prev.next = n.next
	n.next.prev = n.prev
}

// insertFront places a node right after the head (most recently used position).
func (l *LRUCache) insertFront(n *node) {
	n.next = l.head.next
	n.prev = l.head
	l.head.next.prev = n
	l.head.next = n
}

func main() {
	cache := NewLRUCache(3)

	cache.Put(1, 10)
	cache.Put(2, 20)
	cache.Put(3, 30)
	fmt.Printf("Get(1): %d\n", cache.Get(1)) // 10 (moves 1 to front)

	cache.Put(4, 40) // Evicts key 2 (LRU)
	fmt.Printf("Get(2): %d\n", cache.Get(2)) // -1 (evicted)
	fmt.Printf("Get(3): %d\n", cache.Get(3)) // 30
	fmt.Printf("Get(4): %d\n", cache.Get(4)) // 40

	cache.Put(5, 50) // Evicts key 1
	fmt.Printf("Get(1): %d\n", cache.Get(1)) // -1 (evicted)
	fmt.Printf("Get(5): %d\n", cache.Get(5)) // 50
}
